xir/parser.py

```diff
@@ -260,16 +260,30 @@ def obs_stmt(self, args):
             ObservableStmt: object containing statement data
         """
         pref = simplify_math(args[0])
-        terms = args[1]
-        return ObservableStmt(pref, terms, use_floats=self.use_floats)
+        factors = args[1]
+        return ObservableStmt(
+            pref, [ObservableFactor(*factor) for factor in factors], use_floats=self.use_floats
+        )
 
     def obs_group(self, args):
         """Group of observables used to define an observable statement.
 
         Returns:
             list[tuple]: each observable with corresponding wires as tuples
         """
-        return [(args[i], args[i + 1]) for i in range(0, len(args) - 1, 2)]
+        obs_group_list = []
+        args_iterator = iter(args)
+        while True:
+            try:
+                obs = next(args_iterator)
+            except StopIteration:
+                break
+            arg = next(args_iterator)
+            if not is_param(arg):
+                obs_group_list.append((obs, [], arg[1]))
+            else:
+                obs_group_list.append((obs, arg[1], next(args_iterator)[1]))
```

Comment:
This one is completely up to you, but my two cents are that over-using `next()` can result in code that is less readable than using indices. Also, it would be nice to give names to the items in the tuple appended to `obs_group_list`.
```suggestion
        index = 0
        while index < len(args):
           obs, params_or_wires_arg = args[index], args[index + 1]
           
           if is_param(params_or_wires_arg):
               params = params_or_wires_arg[1]
               wires = args[index + 2][1]
               index += 3
           else:
               params = []
               wires = params_or_wires_arg[1]
               index += 2
           
           obs_group_list.append((obs, params, wires))
```